### ISSUE TITLE
Show status of changes on remote and locally (without user action)

### ISSUE THREAD

**Issue by Phillipus** [archi-contribs/archi-modelrepository-plugin#40]
- Provide feedback about what's going on on the local clone vs remote. So this means (for example) changing model icon (in the Collaboration Workspace) when there are uncommitted changes, unpushed commits, available remote changes...
- We could start with providing feedback related to informations/status we already have (unsaved, uncommited, unpushed)
- Then we could add a manual way to do a git fetch and add feedback related to available remote changes
- Then do some git-fetch in the background.

**Comment by Phillipus** [comment]
Some git calls are expensive such as:

- git status (to check for pending commits and pushes)
- git fetch

If there are many local repos in the Collaboration Workspace and we use a background timer or Job to call these it could block the UI.

I thought that for a first implementation we might add a "Refresh" button to the Collaboration Workspace. However, this could be confused with the "Refresh" action on the model...

**Comment by jbsarrodie** [comment]
Maybe we should (again) think about saving a checksum (and size?) of temp.archimate file in the .git folder. It would then be easy to check if the model has been changed (so no need for a costly git status).

I guess that checking if HEAD is ahead of remote/master has no cost.

>I thought that for a first implementation we might add a "Refresh" button to the Collaboration Workspace. However, this could be confused with the "Refresh" action on the model...

Just call it "Check for updates" ;-)

**Comment by Phillipus** [comment]
A checksum might be a good idea.

I've been running some experiments...

IArchiRepository#hasChangesToCommit() can do the git level call, but for this to accurately reflect changes in the git folder, we have to call GraficoModelExporter#exportModel() first to synchronize changes between the temp.archimate file and the xml files in the  .git folder. So it really is an expensive operation to do in a background job.

**Comment by Phillipus** [comment]
I've managed to write a checksum and compare with the temp.archimate file checksum. The issue is one of synchronising when the checksum is written. This could be on a commit, undo commit, abort published changes, and maybe more.

This creates one more problem in the job of keeping things in sync:

temp.archimate file
xml files
checksum file

So, at this stage I'm starting to wonder whether this is a good idea.

**Comment by Phillipus** [comment]
Before we can ascertain the status of commits we need to call GraficoModelExporter#exportModel(). However, this needs the IArchiMateModel to be loaded first. So this is not something that can be run in the background,

**Comment by Phillipus** [comment]
Implemented a background check for unpublished commits and for unpulled commits on the remote. This shows as overlay icons and tooltips in the Collaboration Workspace.

- Could use better overlay icons
- Background job for fetch runs every 20 seconds

**Comment by Phillipus** [comment]
This is not working on GitBlit repos. Under investigation...

**Comment by Phillipus** [comment]
...not GitBlit. It's when a model is pushed from Archi to a git repo. It needs the tracked "master" branch set:

[branch "master"]
	remote = origin
	merge = refs/heads/master

**Comment by jbsarrodie** [comment]
I've tested and it and that fine.

My suggestion for overlay icons: use two arrow on the same bottom corner:
* one from top to bottom to show that there are new things on remote
* on from bottom to top to show that there are things to publish

Something like that:
![overlayicons](https://user-images.githubusercontent.com/5757396/28013836-b4a49012-656a-11e7-9473-39c7a61fb72e.png)

**Comment by Phillipus** [comment]
I tried that icon idea, but you can't see it properly. Typical overlays for a 16x16 icon are 7x8 and can be positioned in each quadrant. I'll try something else...

**Comment by jbsarrodie** [comment]
>I tried that icon idea, but you can't see it properly. Typical overlays for a 16x16 icon are 7x8 and can be positioned in each quadrant

No properly because of their small size or because you can't have two overlay icons on the same quadrant ?

If the last reason, then maybe something like that:
![overlayicons_2](https://user-images.githubusercontent.com/5757396/28016130-589b38d0-6573-11e7-8ed9-d1b6c2acd15d.png)

**Comment by Phillipus** [comment]
It's too small. You can't distinguish it.

How about using the icon itself and coloring certain parts of it:

![overlays](https://user-images.githubusercontent.com/600504/28016649-462fa574-656d-11e7-813d-ce25f3984825.png)

**Comment by jbsarrodie** [comment]
Hmm. Maybe it's time to rethink those icons from the ground up and no more use those eclipse-ish ones.

In such case, we find an icon that can be used to convey the following ideas depending on context (and cumulative):
* Uncommited (local) changes
* Unpublished (local) changes
* Unsynchronised (remote) changes)

**Comment by jbsarrodie** [comment]
>How about using the icon itself and coloring certain parts of it:

I've seen that you've included those new icons on the published release. This seems the better option for the moment, but I would finetune it:
* color the top ball for unsynchronised (remote) changes
* color the bottom left ball for uncommited (local) changes (when we'll be able to detect them)
* color the bottom right ball for unpublished (local) changes

**Comment by Phillipus** [comment]
Yeah, I put it out there unless we have something better. I can juggle the balls easily ;-)

**Comment by Phillipus** [comment]
Latest commits show the 3 colored ball and the Grafico export is triggered everytime the user saves the model. This last thing should be tested in case it slows down the save action.

**Comment by jbsarrodie** [comment]
>And the Grafico export is triggered everytime the user saves the model. This last thing should be tested in case it slows down the save action.

Is it done in the background? If not I can already told you to disable it as I'm working on a model that takes a long time to be exported (approx 30sec).

**Comment by Phillipus** [comment]
It is done in a UIJob so it may be OK. Test it...

(30 seconds? Is that the Grafico export or the save?)

**Comment by jbsarrodie** [comment]
>It is done in a UIJob so it may be OK. Test it...

I've just tested and this seems fine (in fact seems quicker than before, but I'm testing on Linux while I usually use this model on Windows).

>(30 seconds? Is that the Grafico export or the save?)

I think it was mainly the time for JGit to compare HEAD with current tree and find changes to commit.

**Comment by jbsarrodie** [comment]
>I've just tested and this seems fine (in fact seems quicker than before, but I'm testing on Linux while I usually use this model on Windows).

Just tested on Windows and this is really slower slower (12sec) and this blocks UI :-(

I'm opening another issue for that...

**Comment by Phillipus** [comment]
> I'm opening another issue for that...

OK. Let's discuss there... ;-)
